pyomo/network/port.py

```diff
@@ -0,0 +1,512 @@
+            # Name is same, conflicts are prevented by a check in Port.add.
+            # The new var will mirror the original var and have same index set.
+            # This function will be called for both the source and destination
+            # of each arc, but we only need one evar per arc, so check if it
+            # already exists before making a new one.
+            evar = eblock.component(name)
+            if evar is None:
+                evar = Var(index_set)
+                eblock.add_component(name, evar)
+            out_vars.append(evar)
+
+            # Create and potentially initialize split fraction variables.
+            # This function will be called for every Extensive member of this
+            # port, but we only need one splitfrac variable per arc, so check
+            # if it already exists before making a new one.
+            if eblock.component("splitfrac") is None:
+                eblock.splitfrac = Var()
+                splitfracspec = port.get_split_fraction(arc)
+                if splitfracspec is not None:
+                    eblock.splitfrac = splitfracspec[0]
+                    if splitfracspec[1]:
+                        eblock.splitfrac.fix()
+
+            # Create constraint for this member using splitfrac.
+            cname = "%s_split" % name
+            def rule(m, *args):
+                if len(args):
+                    return evar[args] == eblock.splitfrac * var[args]
+                else:
+                    return evar == eblock.splitfrac * var
+            con = Constraint(index_set, rule=rule)
+            eblock.add_component(cname, con)
+
+        if len(out_vars):
+            if write_var_sum:
+                # Create var total sum constraint: var == sum of evars
+                # Need to alphanum port name in case it is indexed.
+                cname = unique_component_name(port_parent, "%s_%s_outsum" %
+                    (alphanum_label_from_name(port.local_name), name))
+                def rule(m, *args):
+                    if len(args):
+                        return sum(evar[args] for evar in out_vars) == var[args]
+                    else:
+                        return sum(evar for evar in out_vars) == var
+                con = Constraint(index_set, rule=rule)
+                port_parent.add_component(cname, con)
+            else:
+                # OR create constraint on splitfrac vars: sum == 1
+                cname = unique_component_name(port_parent,
+                    "%s_frac_sum" % alphanum_label_from_name(port.local_name))
+                con = Constraint(expr=
+                    sum(a.expanded_block.splitfrac for a in port.dests()) == 1)
+                port_parent.add_component(cname, con)
+
+        return out_vars
+
+    # Python 2 compatibility
+    Equality = staticmethod(Equality)
+    Extensive = staticmethod(Extensive)
+    Combine = staticmethod(Combine)
+    Split = staticmethod(Split)
```

Comment:
Ok, so `Equality` --> intensive, `Combine` --> mixer, `Split` --> split fractions... what does `Extensive` do?

Reply:
Extensive is both splitting and mixing. So you can have a single Port with both inputs and outputs, and Extensive will realize this and create a constraint that the sum of the inputs needs to equal the sum of the outputs. Furthermore, the way this is set up, you need to use Extensive if you want your variables to split on one end of an arc and mix on the other, since the rules need to match for connected ports, so you need Extensive on both ends since that is the one that will split for outlets and mix for inlets. I really only split it into 3 functions because it was getting really long as one function, but theoretically a user could use Split as their rule, for example, and it will only split variables and not mix them (not that that makes a whole lot of sense). I also tried to make all of these functions nice and documented and presentable so that they can be used as examples of how to write a custom expansion function.

Reply:
Wait... this setting is by `Port` and not by variable? What if a port has both flow and temperature?

Reply:
The setting is passed as the rule argument to add, so it is specific to each variable. In examples I often talk about ports when I mean a specific variable on the port across multiple ports.